lib/services/client_data_creator.rb

```diff
@@ -0,0 +1,41 @@
+class ClientDataCreator
+  def initialize(client_data, user, attachment_data = [])
+    @client_data = client_data
+    @user = user
+    @attachment_data = attachment_data
+  end
+
+  def run
+    client_data.build_proposal(flow: 'linear', requester: user)
+    client_data.save
+    create_attachments
+    add_public_id
+    proposal
+  end
+
+  private
+
+  attr_reader :client_data, :user, :attachment_data
+
+  def create_attachments
+    attachment_data.each do |file|
+      proposal.attachments.create(user: user, file: file)
+    end
+  end
+
+  def add_public_id
+    proposal.update(public_id: public_id)
+  end
+
+  def public_id
+    if client_data.respond_to?(:fiscal_year)
+      "FY" + client_data.fiscal_year.to_s.rjust(2, "0") + "-#{proposal.id}"
+    else
+      "##{proposal.id}"
+    end
+  end
```

Comment:
I like the idea of encapsulating the proposal creation like this. This `public_id` method is going to get hairy though as we scale out to multiple clients.

Maybe instead delegate to a `build_public_id` method that each `client_data` class could (optionally) implement?